### PR TITLE
8234071: JTable.AUTO_RESIZE_LAST_COLUMN acts like AUTO_RESIZE_ALL_COLUMNS

### PR DESCRIPTION
**Pull request by prsadhuk**
When a JTable is resized with` JTable.setAutoResizeMode` set to ` AUTO_RESIZE_LAST_COLUMN` then it behaves exactly as if I specified `AUTO_RESIZE_ALL_COLUMNS`. 
This is because when `JTable.doLayout` tries to resize the columns, it checks which column to resize by calling `getResizingColumn `and in absence of any column info, it resizes all, so during `setAutoResizeMode` the resizing column needs to be set, which is being done for AUTO_RESIZE_LAST_COLUMN in this fix.
No regression test is provided as it can be easily checked with SwingSet2->JTable(demo)->Autoresize mode (set to "Last Column")

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8234071](https://bugs.openjdk.org/browse/JDK-8234071): JTable.AUTO_RESIZE_LAST_COLUMN acts like AUTO_RESIZE_ALL_COLUMNS (**Bug** - P4)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)
 * [Abhishek Kumar](https://openjdk.org/census#abhiscxk) (@kumarabhi006 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20107/head:pull/20107` \
`$ git checkout pull/20107`

Update a local copy of the PR: \
`$ git checkout pull/20107` \
`$ git pull https://git.openjdk.org/jdk.git pull/20107/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20107`

View PR using the GUI difftool: \
`$ git pr show -t 20107`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20107.diff">https://git.openjdk.org/jdk/pull/20107.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20107#issuecomment-2221343537)

### CODE REVIEW

**Review by prrace — COMMENTED** [review]
Isn't LAST_COLUMN is meant to be different than specifying a particular column ?
Suppose the app specified the resizing column and expects it to be unchanged after changing mode back to a previous mode ?
If the app changes the mode would it really expect now a particular column to be specified ? 
Is a specific test going to be hard ?
No one actually routinely manually tests this in SwingSet, and we don't do integration testing any more.

**Comment by prsadhuk** [comment]
> Isn't LAST_COLUMN is meant to be different than specifying a particular column ? Suppose the app specified the resizing column and expects it to be unchanged after changing mode back to a previous mode ? If the app changes the mode would it really expect now a particular column to be specified ? Is a specific test going to be hard ? No one actually routinely manually tests this in SwingSet, and we don't do integration testing any more.

As per the `JTableHeader.setResizingColumn` spec,
`Application code will not use this method explicitly, it is used internally by the column sizing mechanism`
so app will/should not specify the resizing column directly so I guess it's alright to expect this will not get changed..
I also tested with different autoresize modes and it works fine.

Regarding test, I am not able to think of any automated way to test this, so if it can be manual, then I thought it will be as good to test it out through SwingSet2 since manual tests also is run only during ATR...

**Review by prrace — APPROVED** [review]

**Comment by kumarabhi006** [comment]
I have a doubt for auto resize property, similar behavior is holding true for `AUTO_RESIZE_NEXT_COLUMN` . 
If the user set the mode to `AUTO_RESIZE_NEXT_COLUMN`, all columns are of same width and resized on expanding of container window. As per the spec ( `AUTO_RESIZE_NEXT_COLUMN: Use just the column after the resizing column. This results in the "boundary" or divider between adjacent cells being independently adjustable.`) only the next column should be resized not all the columns.

Is it the correct behavior ? If not, shouldn't `AUTO_RESIZE_NEXT_COLUMN` also needs to be handle along with `AUTO_RESIZE_LAST_COLUMN` ?
**or**
user implementation needs to be changed to handle this mode ?

_Otherwise current fix is working as expected for `AUTO_RESIZE_LAST_COLUMN`._

File: src/java.desktop/share/classes/javax/swing/JTable.java [hunk 1]
```diff
@@ -1266,6 +1266,12 @@ public void setAutoResizeMode(int mode) {
             autoResizeMode = mode;
             resizeAndRepaint();
             if (tableHeader != null) {
+                if (mode == JTable.AUTO_RESIZE_LAST_COLUMN) {
+                    if (columnModel.getColumnCount() > 0) {
+                        tableHeader.setResizingColumn(
+                                columnModel.getColumn(columnModel.getColumnCount() - 1));
+                    }
```
Comment (prsadhuk) [inline src/java.desktop/share/classes/javax/swing/JTable.java:1273]: Check added to guard against the possibility of `DefaultTableColumnModel.getColumn` which according to spec can throw AIOBE
` if columnIndex is out of range: (columnIndex < 0 || columnIndex >= getColumnCount())`

**Comment by prsadhuk** [comment]
> I have a doubt for auto resize property, similar behavior is holding true for `AUTO_RESIZE_NEXT_COLUMN` . If the user set the mode to `AUTO_RESIZE_NEXT_COLUMN`, all columns are of same width and resized on expanding of container window. As per the spec ( `AUTO_RESIZE_NEXT_COLUMN: Use just the column after the resizing column. This results in the "boundary" or divider between adjacent cells being independently adjustable.`) only the next column should be resized not all the columns.
> 
> Is it the correct behavior ? If not, shouldn't `AUTO_RESIZE_NEXT_COLUMN` also needs to be handle along with `AUTO_RESIZE_LAST_COLUMN` ? **or** user implementation needs to be changed to handle this mode ?
> 
> _Otherwise current fix is working as expected for `AUTO_RESIZE_LAST_COLUMN`._

Not sure about the spec you are referring to?
I see https://docs.oracle.com/en/java/javase/22/docs/api/java.desktop/javax/swing/JTable.html#AUTO_RESIZE_NEXT_COLUMN
`When a column is adjusted in the UI, adjust the next column the opposite way.`

It seems to behave similarly with and without the fix, although it may not be what the spec is demanding..It is not similar to LAST_COLUMN spec so cannot be handled along with it..I guess it's a separate issue, if it is one...

**Comment by kumarabhi006** [comment]
> Not sure about the spec you are referring to? I see https://docs.oracle.com/en/java/javase/22/docs/api/java.desktop/javax/swing/JTable.html#AUTO_RESIZE_NEXT_COLUMN `When a column is adjusted in the UI, adjust the next column the opposite way.`

I was referring to the description of `AUTO_RESIZE_NEXT_COLUMN` in [doLayout](https://docs.oracle.com/en/java/javase/22/docs/api/java.desktop/javax/swing/JTable.html#doLayout()) API.

> It seems to behave similarly with and without the fix, although it may not be what the spec is demanding.

Yeah, that's what I was trying to say that changes may be required for `AUTO_RESIZE_NEXT_COLUMN` as it is not as per spec.


It is not similar to LAST_COLUMN spec so cannot be handled along with it..I guess it's a separate issue, if it is one...
Ok.

File: src/java.desktop/share/classes/javax/swing/JTable.java [hunk 2]
```diff
@@ -1266,6 +1266,12 @@ public void setAutoResizeMode(int mode) {
             autoResizeMode = mode;
             resizeAndRepaint();
             if (tableHeader != null) {
+                if (mode == JTable.AUTO_RESIZE_LAST_COLUMN) {
+                    if (columnModel.getColumnCount() > 0) {
```
Comment (kumarabhi006) [inline src/java.desktop/share/classes/javax/swing/JTable.java:1270]: Conditions can be combined

`if (mode == JTable.AUTO_RESIZE_LAST_COLUMN
     && columnModel.getColumnCount() > 0) {`

**Comment by prsadhuk** [comment]
> I was referring to the description of `AUTO_RESIZE_NEXT_COLUMN` in [doLayout](https://docs.oracle.com/en/java/javase/22/docs/api/java.desktop/javax/swing/JTable.html#doLayout()) API.

 "boundary" or divider between adjacent cells being independently adjustable
I guess this is happening...You can resize each column independently by the "divider" between 2 columns (i.e., by dragging the divider left or right, you can see columns can be resize independently)

**Comment by prsadhuk** [comment]
@prrace Need your re-review please..I guess skara is modified to ask for reapproval after PR change (added AIOBE check)

**Review by kumarabhi006 — APPROVED** [review]
Looks good to me.

**Review by prrace — APPROVED** [review]

**Comment by niks** [comment]
Hey @prsadhuk,

we've found a regression in the JTable. After investigating I found out that it was introduced with the fix that you've provided for this issue. Your fix sets the "resizingColumn" of the JTableHeader. That field is documented as "the column being resized. <code>null</code> if not resizing.", which means that is is the column that is resized _right now_, e.g. by the user using the mouse. 

Besides your fix, the method `JTableHeader.setResizingColumn` is only called by `BasicHeaderTableUI` which sets the column actively being resized in a mousePressed event, and clears the value (back to `null`) in the corresponding mouseReleased event.

As a result, your fix does not really work. After the user resized any column using the mouse, the `resizingColunm` is `null` again, so the previous behavior is back (all columns are resized, when you resize the dialog).

And whats worse: It seems like it introduces a regression, where the `TableCellEditor.cancelCellEditing` is called immediately after the editor was added to the table, making it impossible to edit the cell. I need to investigate further, but is is related to the "resizingColumnn", that isn't `null` when the editor should show up.

**Pull request closed without merging**